tc/autotuner/options_cache-inl.h

```diff
@@ -0,0 +1,342 @@
+  }
+}
+
+namespace detail {
+template <typename Backend>
+struct OptionsWithMedianAndRuntimes {
+  Duration median;
+  std::vector<Duration> runtimes;
+  typename Backend::MappingOptionsType mappingOptions;
+};
+
+template <typename Backend>
+std::vector<OptionsWithMedianAndRuntimes<Backend>> sortedOptions(
+    const OptionsCacheKey& key,
+    const std::unordered_multimap<
+        OptionsCacheKey,
+        OptionsCacheValue<Backend>,
+        OptionsCacheKeyHash>& store) {
+  using Options = OptionsWithMedianAndRuntimes<Backend>;
+  std::vector<Options> toSort;
+  if (store.count(key) == 0) {
+    return {};
+  }
+  auto range = store.equal_range(key);
+  for (auto it = range.first; it != range.second; ++it) {
+    if (it->second.runtimes.size() == 0) {
+      throw std::runtime_error("No runtime for cache option");
+    }
+    toSort.push_back(Options{median(it->second.runtimes),
+                             it->second.runtimes,
+                             it->second.mappingOptions});
+  }
+  std::sort(
+      toSort.begin(), toSort.end(), [](const Options& a, const Options& b) {
+        // fun with C++, a.median < b.median does not mix with templates
+        return operator<(a.median, b.median);
+      });
+  return toSort;
+}
+} // namespace detail
+
+template <typename Backend>
+std::vector<typename Backend::MappingOptionsType>
+OptionsCache<Backend>::getTopKOptions(
+    const lang::CanonicalTcString& tc,
+    const std::vector<TensorInfo>& inputs,
+    const std::vector<TensorInfo>& outputs,
+    const std::string& deviceStr,
+    size_t K) const {
+  std::lock_guard<std::mutex> lock(mutex);
+  ++numberAttemptedRetrievals;
+  OptionsCacheKey key{tc, inputs, outputs, deviceStr};
+  auto sorted = detail::sortedOptions<Backend>(key, store_);
+  if (sorted.size() == 0u) {
+    return {};
+  }
+  std::vector<typename Backend::MappingOptionsType> res;
+  res.reserve(K);
+  for (size_t i = 0; i < std::min(K, sorted.size()); ++i) {
+    res.push_back(sorted[i].mappingOptions);
+  }
+  ++numberSuccessfulRetrievals;
+  return res;
+}
+
+template <typename Backend>
+std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash>
+OptionsCache<Backend>::getKeys() const {
+  std::lock_guard<std::mutex> lock(mutex);
+  std::unordered_set<OptionsCacheKey, OptionsCacheKeyHash> keys;
+  for (auto kvp : store_) {
+    keys.emplace(kvp.first);
+  }
+  return keys;
+}
+
+template <typename Backend>
+void OptionsCache<Backend>::pruneKeepTopK(size_t K) {
+  auto keys = getKeys();
+  {
+    std::lock_guard<std::mutex> lock(mutex);
+    for (auto& k : keys) {
+      // this is of course wasteful but real topk is premature optimization atm
+      auto sorted = detail::sortedOptions<Backend>(k, store_);
+      // erase all then reinsert makes it easier to reuse code
+      store_.erase(k);
+      size_t kept = 0;
+      for (auto option : sorted) {
+        if (kept >= K) {
+          break;
+        }
```

Comment:
This looks a bit convoluted.  How about simply
```
for (size_t i = 0; i < std::min(sorted.size(), K); ++i) {
  auto option = sorted[i];
```
Range-based for loops are nice if they improve readability,
but in this case, I would say it's not an improvement.